Improve shared informer notification dispatching

ash2k opened the pull request:
**What this PR does / why we need it**:
Locks and channels don't play well together. This is an attempt to remove locks and only use channels in shared informer. It looks much cleaner to me.

**Release note**:
```release-note
NONE
```

@deads2k @ncdc

ash2k commented:
/sig api-machinery

ash2k commented:
/unassign @dims
/assign @deads2k @ncdc

ncdc commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -466,60 +456,58 @@ type processorListener struct {
+				notification = nil
+			} else {
+				notification = pendingNotifications[0]
+				pendingNotifications[0] = nil
```
Why not just do `pendingNotifications = pendingNotifications[1:]` - why do you need to nil out the first element first?

ash2k replied:
This is to clear the item slot in the slice to remove all references from it to any objects so that GC can clean them up (eventually). Otherwise this slice will "hold" those objects and prevent them from being GCed. See https://github.com/golang/go/wiki/SliceTricks and search for "memory leak". I think the note also applies to this case. Or am I misunderstanding it?

ncdc replied:
What I have seen testing this in the go playground...

- Using `s = s[1:]`
```go
// create a slice of some random type T
s := []*T{x, y, z}
fmt.Printf("%d, %d\n", len(s), cap(s))
// prints out 3, 3

// reset s to elements 1..n
s = s[1:]
fmt.Printf("%d, %d\n", len(s), cap(s))
// prints out 2, 2

c := cap(s)
fmt.Printf("%v\n", s[0:c])
// prints out just 2 elements; the original first element is gone
```

- Using `s = append(s[:0], s[1:]...)
```go
// create a slice of some random type T
s := []*T{x, y, z}
fmt.Printf("%d, %d\n", len(s), cap(s))
// prints out 3, 3

// use append to delete the first element
s = append(s[:0], s[1:]...)
fmt.Printf("%d, %d\n", len(s), cap(s))
// prints out 2, 3

c := cap(s)
fmt.Printf("%v\n", s[0:c])
// prints out 3 elements: y, z, z
```

I think because we are specifically removing the first element using `s = s[1:]` we will be ok w.r.t. GC.

ncdc replied:
Correction, I spoke with some colleagues and it will be safest to nil out the first element, as you're doing.

ash2k replied:
Original `s` and `s` after slicing operation still share the backing array and the first element of that array still contains reference to the object and will not let GC collect it even though this element is not reachable from the resliced `s` and the original slice `s` is gone.

https://play.golang.org/p/gqTJTXdDzU

ncdc replied:
Agreed

ncdc commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -466,60 +456,58 @@ type processorListener struct {
 		case <-stopCh:
 			return
-		case p.nextCh <- notification:
+		case nextCh <- notification:
```
Won't this hotloop here? `run()` is running a for loop to receive from `p.nextCh`, and this function will be able to send `notification` to `nextCh` regardless of if `notification` is nil or non-nil.

ncdc replied:
Oh, wait, you nil out `nextCh`, hmm...

ash2k commented:
ptal @deads2k @ncdc

ncdc commented:
Sorry for letting this slip through. I'll do another pass Monday.

On Sat, Jul 15, 2017 at 3:15 AM Mikhail Mazurskiy <notifications@github.com>
wrote:

> ptal @deads2k <https://github.com/deads2k> @ncdc <https://github.com/ncdc>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/47045#issuecomment-315515969>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYplA1WAnxbQtMAoSyFTIS-a9dRDoks5sOGcVgaJpZM4NxVX0>
> .
>

ncdc commented:
/ok-to-test

ash2k commented:
/retest

ash2k commented:
/retest

ncdc commented:
LGTM. I'd like another reviewer to take a look too.

@kubernetes/sig-api-machinery-pr-reviews @wojtek-t

smarterclayton commented:
Are there performance numbers for this change?  I'd like to make sure this doesn't regress.

ash2k commented:
@smarterclayton I'll add a benchmark for the listener.

ash2k commented:
First number is `concurrencyLevel`:
```
Original:
1: 1000000	      1324 ns/op	     127 B/op	       2 allocs/op
5: 1000000	      1428 ns/op	     134 B/op	       2 allocs/op

PR:
1: 500000	      3139 ns/op	     101 B/op	       1 allocs/op
5: 500000	      3680 ns/op	     105 B/op	       1 allocs/op
```
So it is slower but does 1 less allocation. I don't know how important this slow down is - it probably is not. Also, this is mostly testing Go's concurrency primitives - in practice each listener will be doing something useful which will probably take much more time and amortize the cost.

p.s. I don't understand why my version is slower. Maybe the benchmark is wrong?

ash2k commented:
/test pull-kubernetes-e2e-gce-etcd3

deads2k commented:
> p.s. I don't understand why my version is slower. Maybe the benchmark is wrong?

Logically channels are going to require more lock acquisition.  





> So it is slower but does 1 less allocation. I don't know how important this slow down is

We push a lot of data through this path.  Every mutation to every resource in the entire system, multiple times in some cases.  @smarterclayton @wojtek-t

wojtek-t commented:
Yeah - I also wanted to ask for benchmark but Clayton was first. That's unfortunately what I was expecting, to be honest.

> So it is slower but does 1 less allocation. I don't know how important this slow down is

I think this slow-down is pretty important. It's not 10% slow down - it's almost 3x slow down. As @deads2k already wrote this code path is very busy in pretty much all our components, so I think I'm against this change (though I think we should merge the benchmark as it may be useful in the future too).

ash2k commented:
It will take 3 seconds to push 1 million messages vs 1.3 seconds. How big is this increase relatively to other work that needs to be done to process 1 million messages?

Anyway, if cpu consumption is more important in this case than readability of code I don't mind if this pr will be closed. I didn't expect it to be slower when I opened it so it's ok :)

p.s. I've tried to replace the `pendingNotifications` slice with a ring buffer slice but it is much worse for some reason. Haven't investigated yet. It's here if anyone is curious https://github.com/atlassian/kubernetes/commit/fdb9fd8cd6ff19d18ab163958ed2bfd032c528dc

ash2k commented:
I profiled the benchmark and it is indeed the `select` statement that is causing the slowness. So I removed it from the `add()` method and also simplified the whole thing a bit.

Now it is a proper pipeline: informer starts the processor which starts the listeners, first the `pop()` method and then the `run()` method. `run()` method delivers notifications.

Shutdown is done in reverse order - controller sets the `stopped` flag to prevent addition of new listeners, then stops the processor. Processor closes the `addCh` which causes `pop()` to stop and close the `nextCh` which causes `run()` to stop.

```
1000000	      1451 ns/op	      96 B/op	       1 allocs/op
```

p.s. the additional allocation that is removed here was caused by the `blockingGet` anonymous function that is allocated in the `for` loop in `pop()`.

@deads2k @ncdc

wojtek-t commented:
Those results seem much better - with such a small slow down and 1 allocation less, I think we can merge it. @smarterclayton - thoughts?

ash2k commented:
/retest

liggitt commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -472,80 +478,65 @@ type processorListener struct {
+			if !ok {
+				return
+			}
+			if nextCh == nil { // No notification to pop
```
starting nextCh as nil and having that mean there is no pending notification is hard to follow. seems like it would be simpler to:
1. use p.nextCh everywhere
2. always receive an initial `notification` from addCh and return before getting into the loop if none is available
3. check `if notification == nil` here directly

```
notification, ok := <-p.addCh:
if !ok {
  return
}
```

ash2k replied:
> use p.nextCh everywhere

Then `case p.nextCh <- notification:` will be sending `nil` when there is nothing to send? This `select` case has to be disabled when there is nothing to send. Maybe I don't get what you are proposing?

caesarxuchao replied:
Maybe it's cleaner if you move the manipulation of `nextCh` outside of the select? Something like:
```golang
for {
	if len(pendingNotifications) == 0 {
		nextCh = nil
		notification = nil
	} else {
		nextCh = p.nextCh
		notification = pendingNotifications[0]
	}
	select {
	case nextCh <- notification:
		pendingNotifications[0] = nil
		pendingNotifications = pendingNotifications[1:]
	case notificationToAdd, ok := <-p.addCh:
		if !ok {
			return
		}
		pendingNotifications = append(pendingNotifications, notificationToAdd)
	}
}

```
It's just a matter of perspective, the original code looks ok to me.

ash2k replied:
This looses the optimisation where we don't need to append to the slice: because there is nothing to send in `notification` and `pendingNotifications` is also empty, we can directly put `notificationToAdd` into `notification`. Also approach proposed in the pr only updates variables which must be updated, avoiding unnecessary writes. This is probably not a big deal, but I like it more. Thanks for suggestion anyway!

ash2k commented:
I cannot reproduce the integration test failure locally - I never managed to get integration tests working, this is not the first time. They are always failing to even start. I'm following https://github.com/kubernetes/community/blob/master/contributors/devel/testing.md#integration-tests
Here is what Im' getting:
```shell
make test-integration WHAT=./test/integration/scheduler GOFLAGS="-v"
+++ [0721 10:59:36] Checking etcd is on PATH
/Users/mmazurskiy/code/kubernetes/src/k8s.io/kubernetes/third_party/etcd/etcd
+++ [0721 10:59:36] Starting etcd instance
etcd --advertise-client-urls http://127.0.0.1:2379 --data-dir /var/folders/fh/7zp6kdcj3rgdd_7_nmw439_st0zyrt/T/tmp.KMty7DLi --listen-client-urls http://127.0.0.1:2379 --debug > "/dev/null" 2>/dev/null
Waiting for etcd to come up.
+++ [0721 10:59:37] On try 1, etcd: : http://127.0.0.1:2379
{"action":"set","node":{"key":"/_test","value":"","modifiedIndex":4,"createdIndex":4}}
+++ [0721 10:59:37] Running integration test cases
Running tests for APIVersion: v1,admissionregistration.k8s.io/v1alpha1,admission.k8s.io/v1alpha1,apps/v1beta1,authentication.k8s.io/v1,authentication.k8s.io/v1beta1,authorization.k8s.io/v1,authorization.k8s.io/v1beta1,autoscaling/v1,autoscaling/v2alpha1,batch/v1,batch/v2alpha1,certificates.k8s.io/v1beta1,extensions/v1beta1,imagepolicy.k8s.io/v1alpha1,networking.k8s.io/v1,policy/v1beta1,rbac.authorization.k8s.io/v1beta1,rbac.authorization.k8s.io/v1alpha1,scheduling.k8s.io/v1alpha1,settings.k8s.io/v1alpha1,storage.k8s.io/v1beta1,storage.k8s.io/v1,
+++ [0721 10:59:42] Running tests without code coverage
=== RUN   TestSchedulerExtender
I0721 10:59:50.530022   79314 services.go:35] Network range for service cluster IPs is unspecified. Defaulting to 10.0.0.0/24.
I0721 10:59:50.530432   79314 master.go:182] Node port range unspecified. Defaulting to 30000-32767.
W0721 10:59:50.568217   79314 genericapiserver.go:321] Skipping API autoscaling/v2alpha1 because it has no resources.
W0721 10:59:50.570242   79314 genericapiserver.go:321] Skipping API batch/v2alpha1 because it has no resources.
E0721 10:59:50.617182   79314 controller.go:136] Unable to perform initial Kubernetes service initialization: Endpoints "kubernetes" is invalid: subsets[0].addresses[0].ip: Invalid value: "<nil>": must be a valid IP address, (e.g. 10.9.8.7)
E0721 10:59:50.620908   79314 controller.go:162] unable to sync kubernetes service: Endpoints "kubernetes" is invalid: subsets[0].addresses[0].ip: Invalid value: "<nil>": must be a valid IP address, (e.g. 10.9.8.7)
I0721 10:59:51.602562   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/cluster-admin
I0721 10:59:51.607434   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:discovery
I0721 10:59:51.612679   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:basic-user
I0721 10:59:51.616830   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/admin
I0721 10:59:51.622019   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/edit
I0721 10:59:51.626210   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/view
I0721 10:59:51.629689   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:heapster
I0721 10:59:51.634510   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:node
I0721 10:59:51.639030   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:node-problem-detector
I0721 10:59:51.643089   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:node-proxier
I0721 10:59:51.646633   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:node-bootstrapper
I0721 10:59:51.651696   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:auth-delegator
I0721 10:59:51.655691   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:kube-aggregator
I0721 10:59:51.660592   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:kube-controller-manager
I0721 10:59:51.664463   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:kube-scheduler
I0721 10:59:51.669255   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:kube-dns
I0721 10:59:51.673069   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:persistent-volume-provisioner
I0721 10:59:51.677814   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:attachdetach-controller
I0721 10:59:51.681490   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:cronjob-controller
I0721 10:59:51.687682   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:daemon-set-controller
I0721 10:59:51.692163   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:deployment-controller
I0721 10:59:51.695985   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:disruption-controller
I0721 10:59:51.699856   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:endpoint-controller
I0721 10:59:51.704767   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:generic-garbage-collector
I0721 10:59:51.708969   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:horizontal-pod-autoscaler
I0721 10:59:51.712631   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:job-controller
I0721 10:59:51.716581   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:namespace-controller
I0721 10:59:51.721027   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:node-controller
I0721 10:59:51.726200   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:persistent-volume-binder
I0721 10:59:51.730740   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:pod-garbage-collector
I0721 10:59:51.736285   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:replicaset-controller
I0721 10:59:51.740737   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:replication-controller
I0721 10:59:51.745837   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:resourcequota-controller
I0721 10:59:51.750555   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:route-controller
I0721 10:59:51.755641   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:service-account-controller
I0721 10:59:51.760048   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:service-controller
I0721 10:59:51.763745   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:statefulset-controller
I0721 10:59:51.767657   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:ttl-controller
I0721 10:59:51.772322   79314 storage_rbac.go:178] created clusterrole.rbac.authorization.k8s.io/system:controller:certificate-controller
I0721 10:59:51.776891   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/cluster-admin
I0721 10:59:51.780445   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:discovery
I0721 10:59:51.784554   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:basic-user
I0721 10:59:51.788822   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:node-proxier
I0721 10:59:51.792796   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:kube-controller-manager
I0721 10:59:51.797346   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:kube-dns
I0721 10:59:51.801517   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:kube-scheduler
I0721 10:59:51.806156   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:node
I0721 10:59:51.810413   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:attachdetach-controller
I0721 10:59:51.815239   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:cronjob-controller
I0721 10:59:51.819507   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:daemon-set-controller
I0721 10:59:51.823834   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:deployment-controller
I0721 10:59:51.828194   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:disruption-controller
I0721 10:59:51.832794   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:endpoint-controller
I0721 10:59:51.836747   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:generic-garbage-collector
I0721 10:59:51.841349   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:horizontal-pod-autoscaler
I0721 10:59:51.874026   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:job-controller
I0721 10:59:51.915733   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:namespace-controller
I0721 10:59:51.953852   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:node-controller
I0721 10:59:51.991518   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:persistent-volume-binder
I0721 10:59:52.036848   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:pod-garbage-collector
I0721 10:59:52.072625   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:replicaset-controller
I0721 10:59:52.111480   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:replication-controller
I0721 10:59:52.154360   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:resourcequota-controller
I0721 10:59:52.192011   79314 storage_rbac.go:206] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:route-controller
E0721 10:59:52.275908   79314 status.go:62] apiserver received an error that is not an metav1.Status: rpc error: code = 13 desc = transport is closing
E0721 10:59:52.277165   79314 storage_rbac.go:214] unable to reconcile clusterrolebinding.rbac.authorization.k8s.io/system:controller:service-account-controller: rpc error: code = 13 desc = transport is closing
2017-07-21 11:00:25.591514 I | httptest.Server blocked in Close after 5 seconds, waiting for connections:
  *net.TCPConn 0xc4208a8b30 127.0.0.1:61503 in state active
  *net.TCPConn 0xc4200b0128 127.0.0.1:61505 in state active
E0721 11:00:50.625911   79314 controller.go:149] unable to create required kubernetes system namespace kube-system: Post http://127.0.0.1:61458/api/v1/namespaces: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:50.626474   79314 controller.go:149] unable to create required kubernetes system namespace kube-public: Post http://127.0.0.1:61458/api/v1/namespaces: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.279783   79314 storage_rbac.go:214] unable to reconcile clusterrolebinding.rbac.authorization.k8s.io/system:controller:service-controller: the server cannot complete the requested operation at this time, try again later (get clusterrolebindings.rbac.authorization.k8s.io system:controller:service-controller)
E0721 11:00:52.280147   79314 storage_rbac.go:214] unable to reconcile clusterrolebinding.rbac.authorization.k8s.io/system:controller:statefulset-controller: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/clusterrolebindings/system:controller:statefulset-controller: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.281551   79314 storage_rbac.go:214] unable to reconcile clusterrolebinding.rbac.authorization.k8s.io/system:controller:ttl-controller: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/clusterrolebindings/system:controller:ttl-controller: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.282821   79314 storage_rbac.go:214] unable to reconcile clusterrolebinding.rbac.authorization.k8s.io/system:controller:certificate-controller: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/clusterrolebindings/system:controller:certificate-controller: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.284129   79314 storage_rbac.go:243] unable to reconcile role.rbac.authorization.k8s.io/extension-apiserver-authentication-reader in kube-system: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-system/roles/extension-apiserver-authentication-reader: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.285374   79314 storage_rbac.go:243] unable to reconcile role.rbac.authorization.k8s.io/system:controller:bootstrap-signer in kube-system: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-system/roles/system:controller:bootstrap-signer: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.286663   79314 storage_rbac.go:243] unable to reconcile role.rbac.authorization.k8s.io/system:controller:cloud-provider in kube-system: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-system/roles/system:controller:cloud-provider: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.287740   79314 storage_rbac.go:243] unable to reconcile role.rbac.authorization.k8s.io/system:controller:token-cleaner in kube-system: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-system/roles/system:controller:token-cleaner: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.289194   79314 storage_rbac.go:243] unable to reconcile role.rbac.authorization.k8s.io/system::leader-locking-kube-controller-manager in kube-system: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-system/roles/system::leader-locking-kube-controller-manager: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.290454   79314 storage_rbac.go:243] unable to reconcile role.rbac.authorization.k8s.io/system::leader-locking-kube-scheduler in kube-system: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-system/roles/system::leader-locking-kube-scheduler: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.291733   79314 storage_rbac.go:243] unable to reconcile role.rbac.authorization.k8s.io/system:controller:bootstrap-signer in kube-public: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-public/roles/system:controller:bootstrap-signer: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.292813   79314 storage_rbac.go:275] unable to reconcile rolebinding.rbac.authorization.k8s.io/system::leader-locking-kube-controller-manager in kube-system: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-system/rolebindings/system::leader-locking-kube-controller-manager: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.294050   79314 storage_rbac.go:275] unable to reconcile rolebinding.rbac.authorization.k8s.io/system::leader-locking-kube-scheduler in kube-system: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-system/rolebindings/system::leader-locking-kube-scheduler: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.295224   79314 storage_rbac.go:275] unable to reconcile rolebinding.rbac.authorization.k8s.io/system:controller:bootstrap-signer in kube-system: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-system/rolebindings/system:controller:bootstrap-signer: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.315897   79314 storage_rbac.go:275] unable to reconcile rolebinding.rbac.authorization.k8s.io/system:controller:cloud-provider in kube-system: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-system/rolebindings/system:controller:cloud-provider: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.317068   79314 storage_rbac.go:275] unable to reconcile rolebinding.rbac.authorization.k8s.io/system:controller:token-cleaner in kube-system: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-system/rolebindings/system:controller:token-cleaner: dial tcp 127.0.0.1:61458: getsockopt: connection refused
E0721 11:00:52.318310   79314 storage_rbac.go:275] unable to reconcile rolebinding.rbac.authorization.k8s.io/system:controller:bootstrap-signer in kube-public: Get http://127.0.0.1:61458/apis/rbac.authorization.k8s.io/v1beta1/namespaces/kube-public/rolebindings/system:controller:bootstrap-signer: dial tcp 127.0.0.1:61458: getsockopt: connection refused
F0721 11:01:00.625748   79314 master_utils.go:286] timed out waiting for the condition
goroutine 85 [running]:
k8s.io/kubernetes/vendor/github.com/golang/glog.stacks(0xc42070d000, 0xc4209eab00, 0x57, 0xa3)
        /Users/mmazurskiy/code/kubernetes/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:766 +0xa7
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).output(0x6d96d20, 0xc400000003, 0xc4209eaa50, 0x6b285b5, 0xf, 0x11e, 0x0)
        /Users/mmazurskiy/code/kubernetes/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:717 +0x348
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).printDepth(0x6d96d20, 0xc400000003, 0x1, 0xc42212d878, 0x1, 0x1)
        /Users/mmazurskiy/code/kubernetes/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:646 +0x12d
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).print(0x6d96d20, 0xc400000003, 0xc42212d878, 0x1, 0x1)
        /Users/mmazurskiy/code/kubernetes/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:637 +0x5a
k8s.io/kubernetes/vendor/github.com/golang/glog.Fatal(0xc42212d878, 0x1, 0x1)
        /Users/mmazurskiy/code/kubernetes/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:1125 +0x53
k8s.io/kubernetes/test/integration/framework.startMasterOrDie(0xc420429600, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/mmazurskiy/code/kubernetes/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/framework/master_utils.go:286 +0xa75
k8s.io/kubernetes/test/integration/framework.RunAMaster(0x0, 0x0, 0x0, 0x0)
        /Users/mmazurskiy/code/kubernetes/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/framework/master_utils.go:467 +0x4b
k8s.io/kubernetes/test/integration/scheduler.TestSchedulerExtender(0xc42059b860)
        /Users/mmazurskiy/code/kubernetes/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/scheduler/extender_test.go:291 +0x3f
testing.tRunner(0xc42059b860, 0x39f9fc8)
        /usr/local/Cellar/go/1.8.3/libexec/src/testing/testing.go:657 +0x96
created by testing.(*T).Run
        /usr/local/Cellar/go/1.8.3/libexec/src/testing/testing.go:697 +0x2ca
exit status 255
FAIL    k8s.io/kubernetes/test/integration/scheduler    70.225s
make[1]: *** [test] Error 1
!!! [0721 11:01:00] Call tree:
!!! [0721 11:01:00]  1: hack/make-rules/test-integration.sh:105 runTests(...)
+++ [0721 11:01:00] Cleaning up etcd
+++ [0721 11:01:00] Integration test cleanup complete
make: *** [test-integration] Error 1
```

ash2k commented:
Ok, the test is buggy and is failing (sometimes?) because of a race. Sent a fix #49352.
/kind cleanup

caesarxuchao commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -398,16 +409,21 @@ func (p *sharedProcessor) distribute(obj interface{}, sync bool) {
+	p.listenersLock.RLock()
+	defer p.listenersLock.RUnlock()
+	for _, listener := range p.listeners {
+		close(listener.addCh) // Tell .pop() to stop
```
nit: it's worth mentioning that stopping listener.pod() will stop listener.run(). Otherwise the later comment "Wait for all .pop() and .run() to stop" is confusing.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -472,80 +478,65 @@ type processorListener struct {
+	// pendingNotifications is an unbounded slice that holds all notifications not yet distributed
+	// there is one per listener, but a failing/stalled listener will have infinite pendingNotifications
+	// added until we OOM.
+	// TODO This is no worse that before, since reflectors were backed by unbounded DeltaFIFOs, but
```
s/than/than

ash2k commented:
@caesarxuchao I've added more comments, thanks for good suggestions.

ash2k commented:
/retest

caesarxuchao commented:
/retest

caesarxuchao commented:
LGTM. @deads2k?

smarterclayton commented:
My concerns were addressed, other than making sure we don't regress functionally and keep an eye out for flakes and other logical problems.  I don't know which e2e tests will stress ordering and pathological cases the most though

smarterclayton commented:
Out of curiosity, where's the remaining allocation?  The addNotification{} being converted to an interface?

ash2k commented:
@smarterclayton I believe the source of that allocation is the notifications slice - the reslice operation just creates a new view into the same backing array. But eventually after several appends the array is full and needs to be re-allocated.

I tried playing with ring buffer but I need more time to come up with a proper benchmark that actually shows that there is a benefit (reduced allocations) in using it. Benchmark in this PR does not produce a realistic usage pattern, it just pushes notifications too fast and the ring buffer gets full and needs to be reallocated all the time. However if the consumer is slightly faster than the producer most of the time, the ring buffer should help avoid any allocations. That is a story for another PR.

```bash
go test -c -o cache3.test k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache
./cache3.test -test.v -test.bench ^BenchmarkListener$ -test.run ^$ -test.memprofile=mem3.out
BenchmarkListener-8      1000000              1426 ns/op             109 B/op          1 allocs/op
PASS
go tool pprof --alloc_space cache3.test mem3.out                                           
Entering interactive mode (type "help" for commands)
(pprof) top
103.69MB of 104.19MB total (99.52%)
Dropped 8 nodes (cum <= 0.52MB)
Showing top 10 nodes out of 23 (cum >= 4MB)
      flat  flat%   sum%        cum   cum%
   88.69MB 85.12% 85.12%    88.69MB 85.12%  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*processorListener).pop
   11.50MB 11.04% 96.16%    11.50MB 11.04%  runtime.convT2E
    1.50MB  1.44% 97.60%     4.50MB  4.32%  encoding/json.Unmarshal
       1MB  0.96% 98.56%        1MB  0.96%  k8s.io/kubernetes/vendor/github.com/go-openapi/swag.(*NameProvider).GetJSONNames
       1MB  0.96% 99.52%        1MB  0.96%  runtime.mapassign
         0     0% 99.52%     4.50MB  4.32%  encoding/json.(*decodeState).object
         0     0% 99.52%     4.50MB  4.32%  encoding/json.(*decodeState).unmarshal
         0     0% 99.52%     4.50MB  4.32%  encoding/json.(*decodeState).value
         0     0% 99.52%     4.50MB  4.32%  k8s.io/kubernetes/vendor/github.com/go-openapi/spec.(*Schema).UnmarshalJSON
         0     0% 99.52%        4MB  3.84%  k8s.io/kubernetes/vendor/github.com/go-openapi/spec.MustLoadSwagger20Schema
(pprof) list pop
Total: 104.19MB
ROUTINE ======================== k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*processorListener).(k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.pop)-fm in /Users/mmazurskiy/code/kubernetes/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go
         0    88.69MB (flat, cum) 85.12% of Total
         .          .    376:   p.listenersLock.Lock()
         .          .    377:   defer p.listenersLock.Unlock()
         .          .    378:
         .          .    379:   p.addListenerLocked(listener)
         .          .    380:   p.wg.Start(listener.run)
         .    88.69MB    381:   p.wg.Start(listener.pop)
         .          .    382:}
         .          .    383:
         .          .    384:func (p *sharedProcessor) addListener(listener *processorListener) {
         .          .    385:   p.listenersLock.Lock()
         .          .    386:   defer p.listenersLock.Unlock()
ROUTINE ======================== k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*processorListener).pop in /Users/mmazurskiy/code/kubernetes/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/shared_informer.go
   88.69MB    88.69MB (flat, cum) 85.12% of Total
         .          .    525:                   if nextCh == nil { // No notification to pop
         .          .    526:                           // Optimize the case - skip adding to pendingNotifications
         .          .    527:                           notification = notificationToAdd
         .          .    528:                           nextCh = p.nextCh
         .          .    529:                   } else { // There is already a notification waiting to be dispatched
   88.69MB    88.69MB    530:                           pendingNotifications = append(pendingNotifications, notificationToAdd)
         .          .    531:                   }
         .          .    532:           }
         .          .    533:   }
         .          .    534:}
         .          .    535:
(pprof) ^C
```

deads2k commented:
> LGTM. @deads2k?

beauty is in the eye of the behold, but this looks correct to me.

/approve no-issue

ash2k commented:
@ncdc this one needs lgtm label, please.

ncdc commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -290,6 +288,10 @@ func (s *sharedIndexInformer) AddEventHandlerWithResyncPeriod(handler ResourceEv
 	defer s.startedLock.Unlock()
 
+	if s.stopped {
+		return
```
Do you think this would be worth a log message so people could at least see why their attempt to add an event handler didn't do anything?

ash2k replied:
Done

ncdc reviewed (commented):
1 question about logging, but I don't think it's strictly necessary.

fraenkel commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -472,80 +478,65 @@ type processorListener struct {
 	ret := &processorListener{
-		pendingNotifications:  []interface{}{},
 		nextCh:                make(chan interface{}),
+		addCh:                 make(chan interface{}),
```
Do we really want this unbuffered? I noticed that in the case of just creating the processListener, we created a sync point because we need to hand off the notification to the newly created listener.

ash2k replied:
1. Previous code acquires the lock and then adds the notification to the slice. New code pushes it via a channel. Both have potential to block.
2. The whole purpose of the `pop` method running concurrently is to receive from `addCh` as quickly as possible (or push to `nextCh`).

Benchmark shows that it's not a big deal. I think doing buffering in the channel and then re-buffering into a slice is unnecessary.

fraenkel replied:
It isn't quite the same. The previous code grabbed a lock added it to a slice and unlocked. The new code must wait until you have something pull off the channel.
If you have a backlog of pending notifications, anything calling add will wait until pop actually gets to it.
I am a bit concerned that the Benchmark is just that and may not simulate reality of all the various ways add() is actually invoked.

ash2k replied:
I completely agree that the benchmark is not perfect. I'll happily implement something in addition to it but nothing comes to my mind. If you have any ideas, please share :)

I don't think in practice it matters though - events are coming via the network so delays in order of 0.1-10ms should probably be expected. And we are talking about nanoseconds here.

gmarek commented:
Please squash commits before merging.

ash2k committed:
Improve shared informer notification dispatching

ash2k commented:
@gmarek rebased and squashed.

ncdc commented:
@fraenkel @smarterclayton @deads2k @wojtek-t @gmarek any objections to applying the lgtm label here?

cc @timothysc

smarterclayton commented:
None from me, needs the soak time.

On Wed, Aug 2, 2017 at 9:38 AM, Andy Goldstein <notifications@github.com>
wrote:

> @fraenkel <https://github.com/fraenkel> @smarterclayton
> <https://github.com/smarterclayton> @deads2k <https://github.com/deads2k>
> @wojtek-t <https://github.com/wojtek-t> @gmarek
> <https://github.com/gmarek> any objections to applying the lgtm label
> here?
>
> cc @timothysc <https://github.com/timothysc>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/47045#issuecomment-319674386>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p59GMWaWVl3cpeh2GuYF7vbPqmPRks5sUHu9gaJpZM4NxVX0>
> .
>

deads2k commented:
> @fraenkel @smarterclayton @deads2k @wojtek-t @gmarek any objections to applying the lgtm label here?

Seems reasonable.

ncdc commented:
/retest

ncdc commented:
That's 3 ok's (including me). I'll wait a bit longer, then label

ncdc commented:
/lgtm

ash2k commented:
PR, go away from my dashboard (https://github.com/kubernetes/test-infra/issues/1576)
